src/sleep.rs

```diff
@@ -0,0 +1,110 @@
+//! Driver for light and deep sleep.
+//!
+//! Currently implemented Timer and UART wakeup sources.
+
+use esp_idf_sys::*;
+use std::time::Duration;
+
+#[derive(Debug)]
+pub enum WakeupSource {
+    Timer { dur: Duration },
+    Uart { uart_num: i32, threshold: i32 },
+    // TODO: add more sources
+}
+
+impl WakeupSource {
+    pub fn apply(&self) -> Result<(), EspError> {
+        match *self {
+            WakeupSource::Timer { dur } => {
+                esp!(unsafe { esp_sleep_enable_timer_wakeup(dur.as_micros() as u64) })?;
+            }
+            WakeupSource::Uart {
+                uart_num,
+                threshold,
+            } => {
+                esp!(unsafe { uart_set_wakeup_threshold(uart_num, threshold) })?;
+                esp!(unsafe { esp_sleep_enable_uart_wakeup(uart_num) })?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub trait Sleep {
+    fn add_wakeup_source(&mut self, wakeup: WakeupSource) -> Result<(), EspError>;
+    fn sleep(&self) -> Result<(), EspError>;
+
+    fn apply_wakeup_sources(&self, wakeup_sources: &[WakeupSource]) -> Result<(), EspError> {
```

Comment:
I don't think users are supposed to call this method, so  it does not belong to the trait. However if you remove this method, you can also significantly simplify the whole "sleep" metaphor as well. How about just two simple methods:
* `light_sleep(wakeup: &[WakeupSource])`
* `deep_sleep(wakeup: &[WakeupSource])`

My other concern with this abstraction (and with my newer, simplified variant too) is that you are assuming that the wakeup sources between light and deep sleep are the same. Is that the case, actually?

If they are not, you need a separate `LightSleepWakeupSource` from `DeepSleepWakeupSource`.

Reply:
In terms of a single unified interface over light and deep sleep - YAGNI. And then you'll lose type-safety which is the one sole reason we are doing this. As in - runtime checks for something that can be checked at compile time.

I think we should consider again it makes sense. Are users really be having the need to abstract over light and deep sleep?